Model/Api/ShippingMethods.php

```diff
@@ -441,22 +439,43 @@ function ($el) {
      * Collects shipping options for the quote and received address data
      *
      * @param Quote $quote
-     * @param array $shipping_address
+     * @param array $address_data
      *
      * @return ShippingOptionInterface[]
      */
-    private function getShippingOptions($quote, $shipping_address)
+    private function getShippingOptions($quote, $address_data)
     {
+        if ($quote->isVirtual()) {
+
+            $billingAddress = $quote->getBillingAddress();
+            $billingAddress->addData($address_data);//->save();
+
+            $quote->collectTotals();
+
+            $this->totalsCollector->collectAddressTotals($quote, $billingAddress);
+            $tax_amount = $this->cartHelper->getRoundAmount($billingAddress->getTaxAmount());
+
+            return [
+                $this->shippingOptionInterfaceFactory
+                    ->create()
+                    ->setService('No Shipping Required')
+                    ->setCost(0)
+                    ->setReference(null)
```

Comment:
I suggest you define explicit magic constant string as opposed to using null.

Reply:
done